java-checks/src/main/java/org/sonar/java/checks/TextBlockTabsAndSpacesCheck.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.Collections;
+import java.util.List;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S5664")
+public class TextBlockTabsAndSpacesCheck extends IssuableSubscriptionVisitor {
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.TEXT_BLOCK);
+  }
+
+  private static char indentationCharacter(String[] lines) {
+    for (String line : lines) {
+      if (!line.isEmpty() && isIndentation(line.charAt(0))) {
+        return line.charAt(0);
+      }
+    }
+    // If the string doesn't contain any indentation characters, it doesn't matter what we return
+    return '\0';
+  }
+
+  private static boolean isIndentation(char c) {
+    return c == ' ' || c == '\t';
+  }
+
+  private static boolean containsWrongIndentation(String[] lines) {
+    char indentationCharacter = indentationCharacter(lines);
+    for (String line : lines) {
+      for (int i = 0; i < line.length() && isIndentation(line.charAt(i)); i++) {
+        if (line.charAt(i) != indentationCharacter) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    LiteralTree textBlock = (LiteralTree) tree;
+    String[] lines = textBlock.value().split("\r?\n");
+    if (containsWrongIndentation(lines)) {
+      reportIssue(tree, "Use only spaces or only tabs for indentation");
```

Comment:
I am not exactly sure, but maybe we should report on each violated line rather than on the whole text block. So the issue might be easier to found. Wdyt?

Reply:
I thought about that, but I don't want to create multiple issues for one string literal. What I'd really want is to create a single issue that has the first problematic line as its primary and the others as secondary locations. In fact I wouldn't even want to mark the whole lines, but merely the indentation. Sadly, we don't have fitting a `reportIssue` overload for any of that.

So I'll leave it as-is for now and create a ticket to improve our reporting capabilities here.